aten/src/ATen/Error.h

```diff
@@ -5,14 +5,164 @@
 #include <cstdint>
 #include <cstdio>
 #include <exception>
+#include <sstream>
 #include <stdexcept>
 #include <string>
 #include <type_traits>
+#include <vector>
 
 #include <stdarg.h>
 
+#if !defined(_WIN32)
+#include <cxxabi.h>
+#include <execinfo.h>
+#endif // !defined(_WIN32)
+
 namespace at {
 namespace detail {
+
+// TODO: This backtrace retrieval can be implemented on Windows via the Windows
+// API using `CaptureStackBackTrace` and `SymFromAddr`.
+// https://stackoverflow.com/questions/5693192/win32-backtrace-from-c-code
+// https://stackoverflow.com/questions/26398064/counterpart-to-glibcs-backtrace-and-backtrace-symbols-on-windows
+// https://msdn.microsoft.com/en-us/library/windows/desktop/bb204633%28v=vs.85%29.aspx.
+#if !defined(_WIN32)
+struct FrameInformation {
+  /// If available, the demangled name of the function at this frame, else
+  /// whatever (possibly mangled) name we got from `backtrace()`.
+  std::string function_name;
+  /// This is a number in hexadecimal form (e.g. "0xdead") representing the
+  /// offset into the function's machine code at which the function's body
+  /// starts, i.e. skipping the "prologue" that handles stack manipulation and
+  /// other calling convention things.
+  std::string offset_into_function;
+  /// NOTE: In debugger parlance, the "object file" refers to the ELF file that
+  /// the symbol originates from, i.e. either an executable or a library.
+  std::string object_file;
+};
+
+inline FrameInformation parse_frame_information(
+    const std::string& frame_string) {
+  FrameInformation frame;
+
+  // This is the function name in the CXX ABI mangled format, e.g. something
+  // like _Z1gv. Reference:
+  // https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling
+  std::string mangled_function_name;
+
+#ifdef __GLIBCXX__
+  // In GLIBCXX, `frame_string` follows the pattern
+  // `<object-file>(<mangled-function-name>+<offset-into-function>)
+  // [<return-address>]`
+
+  const auto function_name_start = frame_string.find('(') + 1;
+  const auto offset_start = frame_string.find('+', function_name_start) + 1;
+  const auto offset_end = frame_string.find(')', offset_start);
+  frame.object_file = frame_string.substr(0, function_name_start - 1);
+  frame.offset_into_function =
+      frame_string.substr(offset_start, offset_end - offset_start);
+
+  // NOTE: We don't need to parse the return address because
+  // we already have it from the call to `backtrace()`.
+
+  mangled_function_name = frame_string.substr(
+      function_name_start, (offset_start - 1) - function_name_start);
+#else
```

Comment:
Is there a libcxx specific macro we could check? It would be nicer to avoid assume that everything that's not gcc is clang, and at least have a readable `#error`, or better raise a `#warning` and handle this by returning the whole string without parsing.

Reply:
Looks like `_LIBCPP_VERSION` https://stackoverflow.com/questions/31657499/how-to-detect-stdlib-libc-in-the-preprocessor?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa